src/interrupt/mod.rs

```diff
@@ -0,0 +1,210 @@
+        ty: InterruptSourceType,
+        base: InterruptIndex,
+        count: InterruptIndex,
+    ) -> Result<Arc<Box<dyn InterruptSourceGroup>>>;
+
+    /// Destroy an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object created by
+    /// [create_group()](trait.InterruptManager.html#tymethod.create_group).
+    ///
+    /// Assume the caller takes the responsibility to disable all interrupt sources of the group
+    /// before calling destroy_group(). This assumption helps to simplify InterruptSourceGroup
+    /// implementations.
+    fn destroy_group(&self, group: Arc<Box<dyn InterruptSourceGroup>>) -> Result<()>;
+}
+
+/// Trait to manage a group of interrupt sources for a device.
+///
+/// A device may support several types of interrupts, and each type of interrupt may contain one or
+/// multiple continuous interrupt sources. For example, a PCI device may concurrently support:
+/// * Legacy Irq: exactly one interrupt source.
+/// * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+/// * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+///
+/// PCI MSI interrupts of a device may not be configured individually, and must configured as a
+/// whole block. So all interrupts of the same type of a device are abstracted as an
+/// [InterruptSourceGroup](trait.InterruptSourceGroup.html) object, instead of abstracting each
+/// interrupt source as a distinct InterruptSource.
+#[allow(clippy::len_without_is_empty)]
+#[allow(clippy::trivially_copy_pass_by_ref)]
+pub trait InterruptSourceGroup: Send + Sync {
+    /// Get type of interrupt sources managed by the group.
+    fn interrupt_type(&self) -> InterruptSourceType;
+
+    /// Get number of interrupt sources managed by the group.
+    fn len(&self) -> InterruptIndex;
+
+    /// Get base of the assigned Interrupt Source Identifiers.
+    fn base(&self) -> InterruptIndex;
+
+    /// Get the irqfd to inject interrupts into the guest by using Linux KVM module.
+    fn irqfd(&self, _index: InterruptIndex) -> Option<&EventFd> {
+        None
+    }
+
+    /// Get interrupt flags.
+    fn flags(&self, _index: InterruptIndex) -> u32 {
+        0
+    }
+
+    /// Enable the interrupt sources in the group to generate interrupts.
+    fn enable(&self, configs: &[InterruptSourceConfig]) -> Result<()>;
+
+    /// Disable the interrupt sources in the group to generate interrupts.
+    fn disable(&self) -> Result<()>;
+
+    /// Change the configuration to generate interrupts.
+    ///
+    /// # Arguments
+    /// * index: sub-index into the group.
+    /// * config: configuration data for the interrupt source.
+    fn update(&self, index: InterruptIndex, config: &InterruptSourceConfig) -> Result<()>;
+
+    /// Inject an interrupt into the guest.
+    ///
+    /// If the interrupt has an associated `interrupt_status` register, all bits set in `flag`
+    /// will be atomically ORed into the `interrupt_status` register.
+    fn trigger(&self, index: InterruptIndex, flags: u32) -> Result<()>;
```

Comment:
Yes, it's a tradeoff. For pin-based IRQ, it almost have an associated status register. So the extra parameter `flag` is used to share the common code to manipulate the status register. If we skip the parameter `flags`, each driver will need to implement the code to manipulate status register repeatedly.

Reply:
The code to manipulate the status register is usually one line. So I think that the benefit of having the flags parameter removed is higher than the trouble of handling this outside.

Reply:
The idea is to hide interrupt delivery details from device drivers.
Take virtio-net as an example, the virtio-net driver only cares to trigger an interrupt to notify the guest, and the virtio transport layer decides the way to deliver the interrupt. It doesn't make sense for the virtio-net driver to maintain and manipulate the interrupt status bit flags.
Or it aims to encapsulate interrupt delivery details from device drivers.

Reply:
I agree with encapsulating the information, but I don't think this should be encapsulated at the vm-device level because not all devices need flags. Virtio devices need flags and interrupt_status, so instead of forcing this interface for interrupts on all devices, we can define an interrupt wrapper in vm-virtio. Looking at the implementation of legacy devices for example, it seems like flags is never going to be used. It also looks like the only thing that legacy interrupts and MSI have in common is trigger. I've been playing a bit with separating this interface in multiple specialized interfaces. I hope I'll have a prototype by the end of this week.